src/Negotiation/AbstractNegotiator.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+namespace Negotiation;
+
+abstract class AbstractNegotiator
+{
+    /**
+     * @param string $header     A string containing an `Accept|Accept-*` header.
+     * @param array  $priorities A set of server priorities.
+     *
+     * @return BaseAccept best matching type
+     */
+    public function getBest($header, array $priorities)
+    {
+        if (!$priorities) {
+            throw new \Exception('no priorities given'); 
+        }
+
+        if (!$header) {
+            throw new \Exception('empty header given'); 
+        }
+
+        $headers = self::parseHeader($header);
+
+        $headers = array_map(array($this, 'acceptFactory'), $headers);
+        $priorities  = array_map(array($this, 'acceptFactory'), $priorities);
+
+        $matches = self::findMatches($headers, $priorities);
+
+        $specific_matches = array_reduce($matches, array($this, 'reduce'), array());
+
+        usort($specific_matches, array($this, 'compare'));
+
+        $match = array_shift($specific_matches);
+
+        if ($match === null) {
+            return null;
+        }
+
+        return $priorities[$match->index];
+    }
+
+    /**
+     * @param string $header A string that contains an `Accept*` header.
+     *
+     * @return BaseAccept[]
+     */
+    protected static function parseHeader($header)
+    {
+        $res = preg_match_all('/(?:[^,"]*(?:"[^"]+")?)+[^,"]*/', $header, $matches);
```

Comment:
you should use possessive quantifiers whenever possible to improve the regex by avoiding useless backtracking